### ISSUE TITLE
Evaluation Code

### ISSUE THREAD

**Issue by rwn17** [facebookresearch/vggt#45]
Hi @jytime ,

Thanks for your excellent work! I also evaluate the monocular depth result and it's almost the sota even though it is not specifically trained for that. It's truly amazing!

 was wondering if there are any plans to release the evaluation code (e.g., for pose estimation and MVS) used in the paper? Having access to the official evaluation code would ensure more accurate benchmarking of VGGT's performance, as inappropriate implementations might understate its capabilities.

Bests,
Weining

**Comment by jytime** [comment]
Hi thanks for your interests! Unfortunately I cannot promise about this at this stage. Regarding the evaluation of pose estimation, you can probably check the implementation in our PoseDiffusion and VGGSfM.

Hope I could share something more in the future.

**Comment by jytime** [comment]
Hi @rwn17 ,

When you have gt and pred pointmaps/depthmaps, you naturally have correspondences there, as such grid-based representations have unique mapping.  

We used the simplest Umeyama alignment, and it seems working well to me. Can you talk more about "unsatisfactory"? Does it work but not accurately aligned, or looks totally wrong? If the latter, I guess there is some problems regarding coordinate (e.g., opencv or gl, camera from world or camera to world), or the implementation. Feel free to share some visualisation if necessary.

Best,
Jianyuan

**Comment by lifuguan** [comment]
Hi @jytime , I want to know the number of views input into VGGT during testing and their relationship (e.g., whether there are intervals between them). In my understanding, different numbers and configurations of input views will significantly impact the final metrics.

**Comment by jytime** [comment]
Hi @lifuguan 

These numbers are included in the paper (please correct me if I miss any of them), e.g., 10 frames used for camera pose estimation. For benchmarking, they are just randomly sampled.

**Comment by lifuguan** [comment]
Sorry for the confusion. I was referring to the number of input views in relation to depth and point map accuracy.

**Comment by jytime** [comment]
The depth and point map accuracy is actually chamfer distance, so (in theory) they should not be much affected by the number of input frames I guess? Besides, for example, as discussed in Section 4.3, we use 10 frames for point map evaluation.

**Comment by lifuguan** [comment]
Thanks for the reply!

**Comment by jytime** [comment]
Hi, could you first check if the visualization of the results on RE10K looks correct to you? I’m working on preparing the training code now. I can organize the code for camera pose evaluation after that, but I don’t have an estimated timeline yet.

**Comment by Gaaaavin** [comment]
Hi @rwn17 I'm also trying the same thing. I followed the code in vggsfm but didn't use the function `camera_to_rel_deg` directly. I got results even worse than yours.

I noticed that function has a [`batch_all_pairs`](https://github.com/facebookresearch/vggsfm/blob/e1d9d2eb2b3575525792206fb94b2c749c58dc50/vggsfm/utils/metric.py#L127) step and I'm confused on the purse of that. I'm not sure whether this is the reason why I got worse results. Do you have any idea on this?

It would be really appreciated if @jytime can answer this. Thank you!

**Comment by jytime** [comment]
Hi @Gaaaavin ,

Basically batched_all_pairs will generate N*(N-1)/2 pairs for the given N images.

**Comment by Gaaaavin** [comment]
Thank you for clarification

**Comment by Gaaaavin** [comment]
I also tested the model on the full Realestate10K test set and had similar results to @rwn17 .

The AUC@30 I got is 0.5293

**Comment by jytime** [comment]
Hi @Gaaaavin Do you have results of other methods (e.g., DUST3R) evaluated with your test script? It would help me verify whether the discrepancy comes from different evaluation settings or an issue specific to our model.

**Comment by Ammoniro** [comment]
Hi @jytime Thank you for your fantastic work! I am reproducing the result of MVS and point map estimation. 

I have tried to construct the scene point cloud with depthfusion method as introduced in #MVSNet but the resulting point cloud deviates from the ground-truth(Tested on DTU dataset). As I understand, the predicted camera are in a normalized space which is different from the ground-truth, so I need to align the scene point cloud to GT. 

However, I am struggle with which algorithm I should utilize, as the Umeyama algorithm introduced in paper requires the correspondence of one by one point in GT and the constructed point cloud, it seems only suited for the point map estimation.

Bests, Siyu

**Comment by jytime** [comment]
Hi @Ammoniro ,

For MVS evaluation, I recall I was using the implementation from [GeoMVSNet](https://github.com/doubleZ0108/GeoMVSNet).  Regarding your question — apologies if I misunderstood — but as long as we have depth and camera parameters, we can naturally compute a point map. So it’s not restricted to point map estimation; the same evaluation can be applied for depth estimation as well, right?

However, I think you’re free to use any alignment method that works best for your case. We didn’t explicitly verify this, but it’s very likely you can get better results with a more accurate alignment method than Umeyama.

**Comment by Ammoniro** [comment]
@jytime Thanks for your generous help! I will try more methods.

**Comment by Gaaaavin** [comment]
@jytime Thank you for the suggestion. I'll check the visualization and other methods and update soon.

**Comment by rwn17** [comment]
Hi @jytime ,

I checked some visualizations. Generally, it looks good. But sometimes I find it's not accurate enough. Following is the code I use to get pointcloud 
``` python
  with torch.no_grad():
      aggregated_tokens_list, ps_idx = model.aggregator(images)
      depth_map, depth_conf = model.depth_head(aggregated_tokens_list, images, ps_idx)
      pose_enc = model.camera_head(aggregated_tokens_list)[-1]
      extrinsic, intrinsic = pose_encoding_to_extri_intri(pose_enc, images.shape[-2:])
      point_map_by_unprojection = unproject_depth_map_to_point_map(depth_map.squeeze(0), 
                                              extrinsic.squeeze(0), 
                                              intrinsic.squeeze(0))
```

The following are three examples of large movements.
20d86cff490c0c42
<img width="1253" alt="Image" src="https://github.com/user-attachments/assets/01937665-d04b-46b2-81f7-81a1181aa466" />
<img width="566" alt="Image" src="https://github.com/user-attachments/assets/5c176b15-5c3d-41b6-836c-46a0b7173c29" />

29e0bfbad00f0d5e
<img width="1263" alt="Image" src="https://github.com/user-attachments/assets/c25895f1-6838-4f46-98a9-863b30452143" />
<img width="469" alt="Image" src="https://github.com/user-attachments/assets/1644add7-7fe5-415c-b685-1f30ba22da9c" />

21e794f71e31becb
<img width="1248" alt="Image" src="https://github.com/user-attachments/assets/18024355-2ce8-4b00-8af7-dbb96082b340" />
<img width="586" alt="Image" src="https://github.com/user-attachments/assets/0c7aa273-b898-4017-9d55-d33e96ad6d45" />

Most scenes show good reconstruction quality, except for the last scene with large rotation. I have uploaded the ply file and images to [link](https://drive.google.com/drive/folders/1ztTebNnO7_PwmlLHuwmldFpe7d6QoQqe?usp=sharing).Do you think these reconstructions match the quality you observed in your experiments? Thanks in advance!

**Comment by jytime** [comment]
Hi @rwn17 ,

Thanks for sharing this! It doesn’t look very clean in my own trials. Given that, I’ll clean up the evaluation code and release it for easier evaluation (after the release of the training code).

**Comment by rwn17** [comment]
Hi @jytime , 

Many thanks! I'd really appreciate it if multi-view pose estimation evaluation could be prioritized; the feed-forward version is enough for me.

**Comment by Gaaaavin** [comment]
I did some further experiment and find the way how the 10 input images are chosen can affect results significantly. 

The original 0.5293 AUC@30 I got is by randomly unordered sampling images from the scane:
```
np.random.choice(images.shape[0], 10, replace=False)
```

I get similar result to @rwn17 (AUC@30 = 0.6540) when I uniformly sample images from the scene:
```
np.linspace(0, images.shape[0]-1, 10, dtype=int)
```

However, I believe the random sampling is the same as described in the paper.

@jytime Can you confirm if my random sampling strategy is the same as that in the paper?

@rwn17 I think it would be helpful if you can try random sampling and check whether you get similar results as mine.

**Comment by rwn17** [comment]
Hi @Gaaaavin ,

After using the random sampling, I get similar number 

```
Final evaluation results for re10k:
R@5: 0.9392, R@15: 0.9942, R@30: 0.9996
T@5: 0.2595, T@15: 0.5973, T@30: 0.7945
AUC@30 (all data): 0.5354
```

**Comment by jytime** [comment]
Hi @rwn17 and @@Gaaaavin 

Yes we just used random indexes for sampling. For Co3D, I think we simply used 518x518 by center cropping and resizing.

**Comment by nickle-fang** [comment]
@jytime I appreciate your team's excellent work! But I have a small question about your paper. In Table 10 you compare the pose estimation result on IMC dataset with other methods. [PhotoTourism](https://www.cs.ubc.ca/~kmyi/imw2020/data.html) dataset has 2019, 2020, 2021 and 2022 versions. In 2022 version the test dataset contains all the images but other version only sample 100 images. I wonder which version you choose and how many images do you use to reconstruct.
Thanks a lot!

**Comment by nickle-fang** [comment]
@jytime Sorry for bothering you again! Table 1 in your paper shows the pose estimation result on Re10K and CO3Dv2. I have some questions.
1. The test set in Re10K has about 7.7k videos. How do you sample the video? Do you evaluate all of them?
2. The test set in CO3Dv2 has about 2500 cases. How do you sample them? Do you evaluate all of them?
3. You randomly sample 10 views to estimate poses. I wonder whether the choices of the 10 views will affect the results. Could you provide the 10 indexes of the choice?
4. Could you provide the sample list in Re10K and CO3Dv2 that you choose?

**Comment by rwn17** [comment]
After fixing the numerical issues with torch.inv(), I've achieved an AUC@30 of ~0.74. While below the paper's results, it's around sota. I guess the difference might be due to RE10K's small baseline.

**Comment by jytime** [comment]
Hi @rwn17 ,

Thanks for identifying this! I guess the remaining performance gap is likely due to something similar. 

Btw, @Gaaaavin would the observation of @rwn17 also work for you?

**Comment by jytime** [comment]
Hi @nickle-fang ,

As in VGGSfM, we used IMC 2021, which can be downloaded as below:

wget https://www.cs.ubc.ca/research/kmyi_data/imc2021-public/imc-2021-test-gt-phototourism.tar.gz

For the questions, 

1. In the github repo of posediffusion (https://github.com/facebookresearch/PoseDiffusion), we released the test split for re10k, which contains 1.8k sequences. We randomly sampled frames.
2. For CO3Dv2, we reported the results over all the scenes in the evaluation set. We randomly sampled frames.
3. The indexes are randomly sampled independently for every sequence/scene.

Basically we follow the setting of posediffusion for the evaluation on co3d and re10k. Please check it for more details.

**Comment by jytime** [comment]
Hi @rwn17 ,

May I ask which part did you use torch.inv for, e.g., the inverse of extrinsic matrix? I thought we have provided a function closed_form_inverse_se3 to get the inverse of the extri matrix without numerical problem. Did that help?

**Comment by rwn17** [comment]
> Hi [@rwn17](https://github.com/rwn17) ,
> 
> May I ask which part did you use torch.inv for, e.g., the inverse of extrinsic matrix? I thought we have provided a function closed_form_inverse_se3 to get the inverse of the extri matrix without numerical problem. Did that help?

Hi @jytime, yes, I use torch.inv in my dataloader to inverse the extrinsic matrix instead of using a closed-form inverse. That might be the issue. I also noticed that inverting both the predicted pose and the ground truth pose causes the metric to drop. I'll investigate this in more detail when I have time

**Comment by jytime** [comment]
Hi @rwn17 ,

FYI, the evaluation code of camera pose on Co3D has been uploaded to https://github.com/facebookresearch/vggt/tree/evaluation. I am going to then share the VGGT+BA code, and evaluation code on Re10k (probably also need to complete the training code though).

**Comment by ZhouTimeMachine** [comment]
I use similar setting with the pose estimation of Co3D [(released code)] (https://github.com/facebookresearch/vggt/blob/evaluation/evaluation/test_co3d.py), and I also get AUC@30 of ~0.74 on Re10k, while I reproduce AUC@30 of ~0.88 on Co3D. I guess the performance gap may come from:

1. [test split of Re10k in PoseDiffusion](https://github.com/facebookresearch/PoseDiffusion/blob/main/pose_diffusion/datasets/re10k_test_1800.txt) contains 1832 sequences, but I only have access to 1756 of it, the youtube urls of the rest sequences are unaccessible now. Is there any suggestion to access these data?
2. I randomly sample 10 frames from each sequence, maybe vggt evaluation on Re10k use different sampling strategy from Co3D?
3. Any other suggestion on some details which I might have ignored.

**Comment by rwn17** [comment]
Thanks for the kind and timely support from the author @jytime ! My issue has been well solved and i believe the full evaluation code will be released in the future. I will close this issue.

**Comment by jytime** [comment]
Hello @rwn17 , this might be quite late while I am hoping to share this in case it may help. Is the bug below related to the issue with Re10k?

https://github.com/facebookresearch/vggt/issues/229
